fluent-dom/src/localization.js

```diff
@@ -273,27 +282,24 @@ function messageFromContext(ctx, errors, id, args) {
  */
 function keysFromContext(method, ctx, keys, translations) {
   const messageErrors = [];
-  let hasErrors = false;
+  let missingIds = null;
 
   keys.forEach((key, i) => {
     if (translations[i] !== undefined) {
       return;
     }
 
-    messageErrors.length = 0;
-    const translation = method(ctx, messageErrors, key[0], key[1]);
-
-    if (messageErrors.length === 0 ||
-        !messageErrors.some(e => e instanceof L10nError)) {
-      translations[i] = translation;
+    if (ctx.hasMessage(key[0])) {
+      messageErrors.length = 0;
+      translations[i] = method(ctx, messageErrors, key[0], key[1]);
+      // XXX: Report resolver errors
     } else {
-      hasErrors = true;
-    }
-
-    if (messageErrors.length && typeof console !== "undefined") {
-      messageErrors.forEach(error => console.warn(error));
+      if (missingIds === null) {
+        missingIds = new Set();
```

Comment:
Nit: I wonder if it might be cheaper (and less code) to just create this set upfront rather than check if `missingIds` is null for every translation.

Reply:
My logic here is that this is an error scenario. Even if creating it is cheaper for when there's a single error in 100 strings, it is more expensive if there are zero errors which is 99% of production cases.

I'm optimizing for the scenario where this branch of `if/else` is never hit in production.

Reply:
I understand the motivation. I don't agree with it. Creating a single empty Set per context is cheap and it spares you the trouble of having to check if it's `null` everywhere which is prone to errors. And which, by the way, the code in `formatWithFallback` doesn't do.